server/src/unifyfs_client_rpc.c

```diff
@@ -863,3 +863,68 @@ static void unifyfs_mread_rpc(hg_handle_t handle)
     }
 }
 DEFINE_MARGO_RPC_HANDLER(unifyfs_mread_rpc)
+
+/* Request a list of all the GFIDs the server knows about */
+static void unifyfs_get_gfids_rpc(hg_handle_t handle)
+{
+    int ret = UNIFYFS_SUCCESS;
+    hg_return_t hret;
+
+    /* get input params */
+    unifyfs_get_gfids_in_t* in = malloc(sizeof(*in));
+    if (NULL == in) {
+        ret = ENOMEM;
+    } else {
+        hret = margo_get_input(handle, in);
+        if (hret != HG_SUCCESS) {
+            LOGERR("margo_get_input() failed");
+            ret = UNIFYFS_ERROR_MARGO;
+        } else {
+            client_rpc_req_t* req = malloc(sizeof(client_rpc_req_t));
+            if (NULL == req) {
+                ret = ENOMEM;
+            } else {
+                unifyfs_fops_ctx_t ctx = {
+                    .app_id = in->app_id,
+                    .client_id = in->client_id,
+                };
+
+                // The handler function doesn't actually need the input struct
+                // for anything, so we can free it now.
+                margo_free_input(handle, in);
+
+                req->req_type = UNIFYFS_CLIENT_RPC_GET_GFIDS;
+                req->handle = handle;
+                req->input = NULL;
+                req->bulk_buf = NULL;
+                req->bulk_sz = 0;
+                ret = rm_submit_client_rpc_request(&ctx, req);
+            }
+
+            if (ret != UNIFYFS_SUCCESS) {
+                if (NULL != req) {
+                    free(req);
+                }
+            }
```

Comment:
If ``ret`` is set to an error on line 901, I think the guard here prevents freeing ``req`` even though it may have been allocated successfully.

Reply:
I think the logic here is correct:   If the return value is NOT `UNIFYFS_SUCCESS`, then we need to free the request.

Testing for `NULL` before calling `free()` isn't necessary, though.